README.md
# DaedalusX64

<h1 align="center">
    <a href="https://github.com/DaedalusX64/Daedalus" target="_blank">
        <img height="60%" width="50%" src="pic/ICON0.PNG"><br>
</h1>


[![Discord](https://img.shields.io/badge/Chat-%20Discord-7289da.svg?logo=discord)](https://discord.gg/FrVTpBV)
<a href="https://github.com/DaedalusX64/daedalus/releases" alt="Release: 1.1.8"><img src="https://img.shields.io/badge/Release-1.1.8-%235c913b.svg"></a>
[![Github Language](https://img.shields.io/badge/Language-C++-white?style.svg)](pic/Language.png)
[![GitHub Issues](https://img.shields.io/github/issues/DaedalusX64/Daedalus.svg)](https://github.com/DaedalusX64/daedalus/issues)
[![GitHub stars](https://img.shields.io/github/stars/DaedalusX64/daedalus.svg)](https://github.com/DaedalusX64/daedalus/stargazers)
[![GitHub forks](https://img.shields.io/github/forks/DaedalusX64/daedalus.svg)](https://github.com/DaedalusX64/daedalus/network/members)
<a href="https://github.com/DaedalusX64/daedalus/blob/master/copying.txt" alt="License: GPLv2"><img src="https://img.shields.io/badge/License-GPL%20v2-%23ff554d.svg"></a>

 
<b><u>DaedalusX64</u></b> is a free and open-source Nintendo 64 emulator for Windows, Mac, Linux, PSP, PSVita, PS2 and 3DS with plans to port to among other platforms. Written in C++. It supports most games and has high compatibility.


## Platforms

| Consoles                       | Available               | Latest Version |
| ------------------------------- | ---------------------------- | --------------------- |
| [![pic](pic/PSP.png)](https://github.com/DaedalusX64/daedalus)                          | ✓                       | 1.1.8 |
| [![pic](pic/PSVita.png)](https://github.com/Rinnegatamante/DaedalusX64-vitaGL)                 | ✓              | 0.6 |
| [![pic](pic/PS2.png)](https://github.com/belek666/daedalus)                         | ✓        | 1.1.8 r2 |
| [![pic](pic/3DS.png)](https://github.com/masterfeizz/DaedalusX64-3DS)                              | ✓                  | Beta 0.2 |
| ![pic](pic/PC.png)                              | !               | ! |
| ![pic](pic/macOS.png)                              | !                | ! |
| ![pic](pic/Linux.png)                       | !                   | ! |
 
## Features
 
- Fast emulation
- High compatibility
- Support for PSP TV Mode with 480p output!
- Active support and updates
- Lots of experimental optimizations
- Working savestates
- ARM DynaRec
- Audio support (synchronous and asynchronous)
- Rumble Pak support and Controller Pak support
- Multilanguage support
- [Cheats](Data/Daedalus.cht) Support
- Several more features
 
And many more!

## Screenshots

![pic](pic/1.CBFD.png)
![pic](pic/2.CBFD.png)
![pic](pic/1.SF64.png)
![pic](pic/2.SF64.png)
![pic](pic/SM64.png)
![pic](pic/LZ.png)
![pic](pic/MK.png)

## Compatibility

DaedalusX64 is a High Compatibility Emulator.

[PSP Compatibility List](https://github.com/DaedalusX64/compatibility-list/issues) + 
[DaedalusX64.github.io](https://Daedalusx64.github.io)

- Some roms won't boots.
- Some roms have serious graphical glitches.
- Some roms have random lockups and crashes, or run slowly.
- [SD2DEX](https://github.com/rdx-O/daedalus/blob/3e855a6e458c9c9908dea715aa222ca22bd22a85/Data/S2DEX%20game%20list.txt) Games List
 
## Usage

Install DaedalusX64 to your PSP, download the latest release from the Releases page: 

<h1 align="center">
    <a href="https://github.com/DaedalusX64/daedalus/releases/download/1.1.8/DaedalusX64_1_1_8.zip" target="_blank">
        <img height="20%" width="60%" src="pic/Download.png"></br>
</h1>

##### Next,

Download and Unzip `DaedalusX64.zip` file into `/PSP/GAME/` folder on the memory stick. Place your ROM files in `DaedalusX64/Roms/` and they will automatically appear in Daedalus.

**Main Menu**

When you first run *DaedalusX64* you will be presented with a nice list of your
games. To select a game you will press `D-Pad left/right` then `X` to select a
title, there's various options you can choose from in the menu, every option is
explained. To cycle through options screens use the left and right shoulder
buttons.
 
**Note** : If the release is a .zip file with a folder containing an `EBOOT.PBP` file when extracted, simply drag and drop the extracted folder into `/PSP/GAME/`.
DaedalusX64 supports most rom formats ```.v64```, ```.z64```, ```.rom```

## Controls

Default controls :

| N64                             | PSP                                   |
| ------------------------------- | ---------------------------- |
| Start                           | Start                        |
| Analogue Stick                  | Analogue Stick               |
| Dpad                            | ○ (Circle) +  Dpad          |
| A                               | Ｘ (Cross)                   |
| B                               | □ (Square)                  |
| Z                               | △ (Triangle)                |
| L Trigger                       | L Trigger                    |
| R Trigger                       | R Trigger                    |
| C buttons                       | Dpad (Circle unpressed)      |
| Menu                            | Home                         |  

## Building for Windows 

1) Clone and open the repo in Visual studio 2019

2) Build All

## Building for Linux and Mac

1) Clone this repo 

2) Open a terminal and run `build_daedalus.sh` with the argument for the os you would like to build for 
Example: ```./build_daedalus.sh MAC```

## Optimal Settings

<Details>

### :: Preferences ::

![Screenshot](pic/picp.png)

**Texture Update Check** : Every Frame (This makes sure the game's textures runs smoothly and that no visual glitches occur, may want to disable for more fps but significant drop in quality)

**Frameskip** : Disabled (Most games run fine with this disabled. If you do need to enable it, it will make the game more choppy but will grant you an FPS boost)

**Zoom** : 100% (Not much explanation needed)

**Limited Framerate** : Full speed (This will make sure that your FPS is not limited so that you can get the maximum FPS you can)

**Dynamic Recompilation** : Enabled (This makes the system recompile some part of a program during execution, making the emulator more efficient and optimised)

**High level Emulation** : Enabled (Optimises the game even more)

**Audio** : Synchronous (Some games may just have bad/laggy audio so you might want to turn this off if the audio is still choppy)

**Controller** : Default (Unless you have a custom layout in which you should use that)

### :: Advanced Options ::

**Dynarec Memory Optimisation** : Enabled

**Dynarec Loop Optimisation** : Enabled (NOTE: Very unstable - if your game has crashes, this is probably why)

**Dynarec Doubles Optimisation** : Enabled

**Clean Scene** : Disabled

**Clear N64 Depth Buffer** : Disabled

**Double Display Lists** : Enabled

**Video Rate Match** : Yes

**Audio Rate Match** : No

**Fog Emulation** : Disabled

**NOTE** : When the sound is activated, the emulation speed tends to decrease and the game has errors, which is why it is deactivated.

Here we put the **Previews** of most of the Nintendo 64 games, simply download this link and the `Resources` folder of the download, place it inside your emulator.

</Details>

## Changes History

<Details>

**1.1.8**

* New graphical interface and small images of the covers of the games have been added, which work properly with large lists of roms.
* Changes have been made to asynchronous audio making the use of the PSP's Media Engine more stable .
* The PRX in charge of the Media Engine processes has been updated , improving its performance.
* Now you have a new logo and background image the EBOOT.PBP.
* Added a new speed hack that improves the performance of some games.
* Now it detects if the emulator is running on a PS Vita.
* The code has been updated to the C ++ 11 standard .
* General code cleaning

**1.1.7**

* New UI and small preview images: Working properly with large ROM lists
* Changes to Asynchronous audio: No longer waits for the Media Engine CPU to complete the last job and run on the main PSP CPU!
* New EBOOT logo and PIC1
* Rebased with the latest version from strmnnrmn
* C11 Standard variables
* General code cleanup and more!

Wally4000 : New UI, Code cleanup, Re-basing
MotoLegacy : New icons
z2442 : ME work, Code cleanup

**1.1.6**

* Updated our toolchain which was provided by top-sekret, this provides GCC 8 and oddly has improved performance of Daedalus
* Make some changes to the job manager to improve audio performance, Audio is very stable now.
* Enable more fast memory copying

**1.1.5**

* Move to a modern make system - Daedalus was left in 2009 with gyp as it's main compiler but that lead to trouble and the PSP build did not work with that. I've moved all the source files and managed to get it compiing in CMake which to me is by far a more flexible system than gyp. This will allow us to finally start to look at making the macOS / Linux and Windows versions working again.
If you've built the source in the past. (Not sure if StrmnNrmn would approve but like make did to him, gyp makes me want to tear my eyes out. :P
There are still some things that need to be moved, like the PSP PRX sources but that can wait as it's not important.
* Clean up the Source - Still work in progress but is happening as there appear to be other projects in the pipeline. The source has lost some weight and likely will continue to do so until it's nice and clean

**1.1.4**

* improved stability and performance
* minor changes to the dynarec
* updated some UI aspects to remove graphical bugs

-Notes
* If using a ps vita please delete the MediaEngine.prx form the DaedalusX64 folder. This will prevent issues when running audio in async mode. The vita does not have and does not emulation the psps co processor.
* Preview images have been removed from the binary file since it is to large to for github. If you would like them please down load them from this repo and place them in \DaedalusX64\Resources\Preview

**1.1.3**

* Further improvements to code optimization including more efficient maths and faster code for the Media Engine bring even further performance improvements.

**1.1.2**

* This release has been updated to compile on the latest PSP SDK. Minor changes to the code and makefile.

**1.1.1**

* Compiled with a small makefile change.

</Details>

## More Information
 
###### [DaedalusX64/Wiki](https://github.com/DaedalusX64/daedalus/wiki) Check *DaedalusX64* ports page for other consoles with information.


<h6>Join our Discord server to talk to other Daedalus users and the developers!</h6>
<h1 align="center">
    <a href="https://discord.gg/FrVTpBV" target="_blank">
        <img height="49%" width="58%" src="pic/DX64.png"><br>
    <a href="https://discord.gg/YTwTRy2" target="_blank">
        <img height="50%" width="58%" src="pic/PHC.png"><br>
</h1>
<p align="center">
 
## Credits

|        |                                                            |
| ----------------- | ---------------------------------------------------------- |
| **kreationz, salvy6735, Corn, Chilly Willy**       | Original DaedalusX64 code     |
| **Wally4000**             | Optimizations, improvements and ports |
| **z2442** | Compilation improvements and updating, optimizations|
| **mrneo240** | Optimizations, compilation help |
| **TheMrIron2** | Optimizations, wiki maintenance |
| **MotoLegacy** | Help on Icon and improvements |

Thank you for everyone who has supported us along the way. The realise would not have been possible without your help!.

### Thanks

[Thanks](pic/Thanks.txt)


pic/Changelog.txt
DaedalusX64 Beta 2 - 1 March 2009
----------------------
[+] Added exception handler to emulation (Chilly Willy)
[+] Made UI selection wrap at top and bottom (Chilly Willy)
[+] Cleaned up the ROM preferences (Chilly Willy)
[+] ME Audio uses CPU event (CPU event system made thread-safe) (Chilly Willy)
[+] ME Audio made conditional (Async = ME, Sync = Old Audio) (Chilly Willy)
[+] Several Custom Blends Added (Kreationz, ShinyDude100, Wally)
[+] Rewrote microcode detection to fix a couple of bugs (StrmnNrmn)
[+] Added support for LOAD_UCODE, fixes a number of rendering glitches (StrmnNrmn)
[+] Speed hack detection (howard0su)
[+] Added JPEG support (howard0su)
[!] Fixed Texture Update Bug causing ghosting in MarioKart (Kreationz)
[!] Fixed an assert causing some homebrew not to run (StrmnNrmn)
[!] Added support for FlashRam and Eeprom16k save type (howard0su)
[!] Reworked saving/loading system (howard0su)
[^] OSEHLE Enabled with several OSHLE commands (Small speed up)(howard0su)
[^] Improved Audio ME code (Faster with Audio on) (Chilly Willy)
[^] Use VFPU to handle more transform and lighting (StrmnNrmn)
[^] Various rendering optimisations (StrmnNrmn)
[^] Various ucodes, many games show up correctly now (Wally)
[^] Corrected many savetypes in rom.ini (Wally)
[~] Various code cleanup (howard0su)


DaedalusX64 Beta 1 - 08 November 2008 (Changes from R14 Beta)
----------------------
[+] TV Out Support (Chilly Willy)
[+] ME Audio Supported (Small speed-up with sound on.) (Chilly Willy)
[+] Increased audio buffers for better sound quality (Kreationz)
[+] Changed logo and graphics to represent new name (graphic by SynGamer)
[!] Fixed screen shots (Chilly Willy)
[!] Fixed an OoT custom clend Mode (Wally)
[!] Fixed Shaking using double display lists (GoldenEye 007, others...) (idea Wally, code Kreationz)
[!] Pause Menu Reset not working properly (Kreationz)
[!] Fixed FPS Display and Framerate limiter (Kreationz)
[!] Optimized code generation for fragments which branch to themselves is now optional
Fixes games that broke from R12-13 (Kreationz)
[^] Improved thread communications in job manager (Hlide)

R13 - 04 November 2007
----------------------

[+] Savestate support.
[^] Improved dynarec load/store code generation to remove a branch and a couple of instructions.
[^] Dynarec now generates much better code for fragments which branch to themselves.
[^] Dynarec load/stores through the stack pointer are now much more efficient.
[^] Avoid invalidating the instruction cache where possible.
[^] Simplified CPU event handling, making it slightly more efficient.

R12 - 26 June 2007
------------------

[!] Fixed issue preventing Goldeneye from being loaded.
[!] Fixed dynarec for Goldeneye.
[!] Fixed dynarec for Super Smash Bros.
[!] Fix various texturing issues with 4bpp and small or non power-of-2 textures.
[!] Fix TexRect instructions with negative s/t components.
[!] Fixed the HUD in Mario 64 (broken in R11.)
[!] Fixed lights in F3DEX2 microcodes.
[+] Correctly implement instruction fetch exceptions, improving compatibility.
[+] Improved floating point compatibility.
[+] Correctly handle mask_s/mask_t tile values.
[+] Implemented a few custom blend modes.
[+] Screenshots just cover visible viewport.

R11 - 06 May 2007
-----------------

[*] Large overhaul of rom settings and preferences. All settings should now be persistant.
[+] Native support for palettised textures for faster rendering and less memory usage.
[+] Provided a fixed size pool for all texture memory, to limit memory usage.
[+] Expansion Pak enabled by default.
[!] Updated rom info database to include save mechanism for many roms.
[!] Fix memory leak in texture cache.
[!] Fixed conversion of RGBA/32bpp textures with odd alignments.
[!] Fix crash which occurred when recolouring textures in low memory situations.
[^] Large rewrite of texture cache to reduce memory usage and improve performance.
[^] Various optimisations to the way textures are looked up in the cache and installed.
[^] Optimise offsetting/scaling of texture coordinates.
[^] Improve performance of dynarec fragment cache and reduce memory fragmentation.

R10 - 02 April 2007
-------------------

[+] Added frameskip option.
[+] Added framerate limiting option.
[+] Added adjustable stick deadzone.
[+] Allow pause menu to be activated even when rom emulation has hung.
[!] Fix crash when no roms found, added explanatory message.
[!] Fix flickering when small number of roms found.
[^] Fix texture hash frequency check.
[^] Finally got code compiling with -O3 optimisation flag in GCC.
[^] Use VFPU for InvSqrt (thanks hlide!)
[^] Improved VFPU code for clipping and tnl (thanks Raphael!)
[^] Improved inlining of AddTri.
[^] Reduced time spent searching for overridden blend modes.
[~] Remove debug console in public release builds (smaller footprint and slight speedup.)

R9 - 22 February 2007
---------------------

[+] Added support for RGBA 4444 and 5551 textures
[!] Tidied up all the texture conversion code, fixing a few bugs in the process
[!] Fixed the width/height of FillRect calls in 1 and 2 cycle mode
[!] Fixed a blending bug
[^] Use 16-bit textures on the PSP to represent 16-bit N64 textures.
[+] Added mirrored texture support
[!] Fixed a LoadTile bug, allowing a couple of hacks to be removed
[+] Added some new blend modes for various roms
[!] Fixed the Tri2 command for F3DLX microcodes
[!] Fixed a bug in busy-wait detection
[!] Fixed a few dynarec stability issues (relating to exceptions occuring mid-trace)
[+] Added audio support
[+] Added the ability to dump textures (developer builds only at the moment)
[!] Fixed screenshots. Again.
[+] Implemented cmp.s, cvt.s, cvt.w, mtc1, mfc1, bc1f, bc1t, j, cfc1, ctc1,
daddu, trunc.w.s, bc1t, bc1f, bc1tl, bcifl, bnel, beql, blezl, bgtzl,
bltzl, blezl in dynarec
[^] Avoid setting the branch delay flag and current PC in generated dynarec code unless necessary
[^] Much better memory access handling in dynamically recompiled code

Tools/roms.py
import zlib
import zipfile
import glob
import os
import struct
import re
import hashlib

class Stats:
  crcs = {}
  microcodes = {}

  def __init__(self):
    self.crcs = {}
    self.microcodes = {}
    
  def addMicrocode(self, microcode, rom_name):
    if not self.microcodes.has_key(microcode):
      self.microcodes[microcode] = []
    self.microcodes[microcode].append(rom_name)
      
  def printMicrocodes(self):
    print '\n\n\n******* Microcodes *************\n\n'
    for microcode in sorted(self.microcodes):
      print microcode
      for rom_name in self.microcodes[microcode]:
        print rom_name
      print '' 
      print ''

    print ''
    print ''
    for microcode in sorted(self.microcodes):
      print microcode

def getCountryNameFromCountryID(country):
  countries = { '7' : "Beta",
                'A' : "NTSC",
                'D' : "Germany",
                'E' : "USA",
                'F' : "France",
                'I' : "Italy",
                'J' : "Japan",
                'P' : "Europe",
                'S' : "Spain",
                'U' : "Australia",
                'X' : "PAL",
                'Y' : "PAL" }
  if country in countries:
    return countries[country]
  return 'blank'

def getIconUrlFromCountryID(country):
  countries = { '7' : "unknown.png",
                'A' : "unknown.png",
                'D' : "de.png",
                'E' : "us.png",
                'F' : "fr.png",
                'I' : "it.png",
                'J' : "jp.png",
                'P' : "europeanunion.png",
                'S' : "es.png",
                'U' : "au.png",
                'X' : "unknown.png",
                'Y' : "unknown.png" }
                
  if country in countries:
    url = countries[country]
  else: url = 'unknown.png'

  return '/images/flags/' + url

def countryFromRomId(rom_id):
  m = re.search('{([a-fA-F0-9]+)-([a-fA-F0-9][a-fA-F0-9])}', rom_id)
  if not m: return '?'
  return chr(int(m.group(2), 16))

class RomDetails:
  data = {}
  rom_id = ''
  rom_name = ''
  cic_type = ''
  sha1_hash = ''
  
  def __init__(self, rom_id):
    self.data = {}
    self.rom_id = rom_id

  def getName(self):
    return self.rom_name
    
  def getSha1Hash(self):
    return self.sha1_hash
    
  def setSha1Hash(self, sha1_hash):
    self.sha1_hash = sha1_hash
    
  def setCicType(self, cic_type):
    self.cic_type = cic_type
    
  def __str__(self):
    return str(self.data)


def addRom(rom_id, attributes):

  if attributes.get('Comment') == 'Unknown':
    del attributes['Comment']

  current_rom = RomDetails(rom_id)
  current_rom.rom_name = attributes.get('Name')
  current_rom.rom_country = countryFromRomId(rom_id)
  current_rom.expansionPakUsage = attributes.get('ExpansionPakUsage')
  current_rom.preview = attributes.get('Preview')
  current_rom.saveType = attributes.get('SaveType')
  return current_rom


def parseRomsIni(lines):
  roms = []
  current_id = ''
  attributes = {}
  
  for line in lines:
    line = line.rstrip()
    if not line:
      continue
    elif line.startswith('{'):
      if current_id:
        roms.append(addRom(current_id, attributes))
      current_id = line
      attributes = {}
    elif current_id:
      vals = line.split('=')
      if len(vals) == 2:
        attributes[vals[0]] = vals[1]
      else:
        print 'invalid line ', line
    
  if current_id:
    roms.append(addRom(current_id, attributes))
  return roms


def parseRomsInifile(filename):
  f = open(filename, 'r')
  roms = parseRomsIni(f.readlines())
  f.close()
  return roms

def writeRomsIni(filename, roms):
  f = open(filename, 'w')
  
  # Sort roms based on Name field
  roms.sort(lambda x, y: cmp(x.getName(),y.getName())) 
    
  for rom in roms:
    f.write(rom.rom_id + '\n')
    f.write('Name=%s\n' % rom.getName())
    for kv in sorted(rom.data):
      f.write(kv + '=' + rom.data[kv] + '\n')
    if rom.cic_type:
      f.write('CicType=%s\n' % rom.cic_type)
    f.write('\n')

  f.close()

def all_files(pattern, search_path, pathsep=os.pathsep):
  for path in search_path.split(pathsep):
    for match in glob.glob(os.path.join(path, pattern)):
      yield match



def isRomHeader(buffer):
  magic = map(ord, buffer[:4])
  return magic in [[0x80, 0x37, 0x12, 0x40],    # 'Correct'
                    [0x40, 0x12, 0x37, 0x80],
                    [0x37, 0x80, 0x40, 0x12],
                    [0x12, 0x40, 0x80, 0x37]]
                    
                    
def fudgeByteswap(buffer):
  r = list(buffer)
  for i in range(0,len(buffer),4):
    r[i+0] = buffer[i+3]
    r[i+1] = buffer[i+2]
    r[i+2] = buffer[i+1]
    r[i+3] = buffer[i+0]
  return ''.join(r)
                
def correctByteswap(buffer):
  magic = map(ord, buffer[:4])
  if magic == [0x80, 0x37, 0x12, 0x40]: 
    return buffer
    
  r = list(buffer)
    
  if magic == [0x40, 0x12, 0x37, 0x80]:
    for i in range(0,len(buffer),4):
      r[i+0] = buffer[i+3]
      r[i+1] = buffer[i+2]
      r[i+2] = buffer[i+1]
      r[i+3] = buffer[i+0]
  elif magic == [0x37, 0x80, 0x40, 0x12]:
    for i in range(0,len(buffer),4):
      r[i+0] = buffer[i+1]
      r[i+1] = buffer[i+0]
      r[i+2] = buffer[i+3]
      r[i+3] = buffer[i+2]
  elif magic == [0x12, 0x40, 0x80, 0x37]:
    for i in range(0,len(buffer),4):
      r[i+0] = buffer[i+2]
      r[i+1] = buffer[i+3]
      r[i+2] = buffer[i+0]
      r[i+3] = buffer[i+1]

  return ''.join(r)

def find_rom(roms, rom_id):
  for rom in roms:
    if rom.rom_id == rom_id:
      return rom
  return None
  
def getCicNameFromCrc(crc):
  name_map = { 0x6170a4a1 : 'CIC-6101',
               0x90bb6cb5 : 'CIC-6102',
               0x0b050ee0 : 'CIC-6103',
               0x009e9ea3 : 'CIC-6104',
               0x98bc2c86 : 'CIC-6105',
               0xacc8580a : 'CIC-6106' } 
  if crc in name_map:
    return name_map[crc]
  return '?'

def findMicrocodes(buffer):
  codes = {}
  idx = buffer.find('RSP')
  while idx != -1:
    has_space = False
    for i in range(idx, idx+200):
      c = ord(buffer[i])
      if c > 128 or c == ord('=') or c == ord('@'):
        break
      if c < 32:
        microcode = buffer[idx:i]
        if len(microcode) > 8 and has_space:
          if not codes.has_key(microcode):
            codes[microcode] = 1
        break
      elif c == ord(' '):
        has_space = True
    idx = buffer.find('RSP', idx+1)
  return codes.keys()

def find_roms_in_zipfile(zip_filename, roms, stats):
  z = zipfile.ZipFile(zip_filename, 'r')
  
  for filename in z.namelist():
    f = z.open(filename, 'r')
    if not f: continue

    buffer = f.read()
    processBuffer(filename, buffer)
  
def processFile(filename, roms, stats):
		buffer = open(filename,'rb').read()
		processBuffer(filename, buffer)
    
def processBuffer(filename, buffer):

    if len(buffer) < header_size+bootcode_size: return
    if not isRomHeader(buffer): return
    
    
    len_mod = len(buffer) % 4
    if len_mod > 0:
      print '%s / %s not multiple of 4 - %d' % (zip_filename, filename, len(buffer))
      padding = chr(0) * (4 - len_mod)
      buffer = buffer + padding
      return
    
    buffer = correctByteswap(buffer)
    
    header_buf = buffer[:header_size]
    bootcode_buf = buffer[header_size:game_offset]
    
    fields = list(struct.unpack(header_string, header_buf))
       
    #crc = long(zlib.crc32(bootcode_buf) & 0xffffffff)
    #new_count = 1
    #if stats.crcs.has_key(crc): new_count = stats.crcs[crc] + 1
    #stats.crcs[crc] = new_count
        
    rom_name = fields[11].rstrip()
    rom_id = '{%08x%08x-%02x}' % (fields[7], fields[8], fields[17])
    
    rom = find_rom(roms, rom_id)
    if rom:
      #rom.setCicType(getCicNameFromCrc(crc))
      #if not rom.getSha1Hash():
      #  sha1_hash = hashlib.sha1(buffer).hexdigest()
      #  print 'sha1 - %s' % sha1_hash
      #  rom.setSha1Hash(sha1_hash)
      print '%s - "%s"' % (filename, rom.getName())

      for s in findMicrocodes(buffer): 
        stats.addMicrocode(s, rom.getName())
        print '  ' + s
    else:
      print '%s - %s - %s' % (filename, rom_name, rom_id)
    
  
#struct ROMHeader
#{
#	u8		x1, x2, x3, x4;   0
#	u32		ClockRate;        4
#	u32		BootAddress;      5
#	u32		Release;          6
#	u32		CRC1;             7
#	u32		CRC2;             8
#	u32		Unknown0;         9
#	u32		Unknown1;         10
#	s8		Name[20];         11
#	u32		Unknown2;         12
#	u16		Unknown3;         13
#	u8		Unknown4;         14
#	u8		Manufacturer;     15
#	u16		CartID;           16
#	s8		CountryID;        17
#	u8		Unknown5;         18
#};
header_string = 'BBBBLLLLLLL20sLHBBHbB'
header_size = struct.calcsize(header_string)
game_offset = 0x1000
bootcode_size = 0x1000 - 0x40

    
stats = Stats()
roms = parseRomsInifile('roms.ini')

count = 0
for match in all_files('*.z64', r'm:\rom_dump'):
  #print match
  #find_roms_in_zipfile(match, roms, stats)
  processFile(match, roms, stats)
  count = count + 1
  #if count >= 30: break
stats.printMicrocodes()
#for crc in sorted(crcs): print 'CRC %s -> %s' % (getCicNameFromCrc(crc), crcs[crc])

writeRomsIni('roms2.ini', roms)
